lib/nimble_options.ex

```diff
@@ -440,6 +442,48 @@ defmodule NimbleOptions do
     end
   end
 
+  defp validate_type({:list, type}, key, values) when is_list(values) do
```

Comment:
This implementation returns an unexpected return value, which is `{:error, custom_errors}`. This breaks the API. Honestly, I'm starting to think that supporting this is more work than gain and supporting it in a solid and nice way is not that straigthforward. I'm tempted to not include this feature. Thoughts @josevalim?

Reply:
My ¢2. I needed this feature (list of lists) and I went with 
```elixir
{:custom, Checker, :list,
  [{Checker, :list, [{Kernel, :is_atom, []}]}]}
```
with the straightforward implementation of `Checker.list/1` as
```elixir
@spec list([any()], (any() -> boolean)) :: {:ok, [any()]} | {:error: binary()}
def list(values, checker) do
  values
  |> Enum.all?(checker)
  |> if(
    do: {:ok, values},
    else: {:error, "Expected list"}
  )
end
```
```

Reply:
It works pretty well for simple values but it's cumbersome to support complex values like custom functions because they must return `{:ok, value}` or `{:error, message}`, which means there's no straightforward way to output all errors from those custom functions without breaking the current API:

```elixir
      schema = [ports: [type: {:list, {:custom, __MODULE__, :validate_port, []}}]]
      opts = [ports: [:not_a_port, 22, "80", __MODULE__]]

      assert NimbleOptions.validate(opts, schema) ==
               {:error,
                [
                  "expected an integer or a string, got: :not_a_port",
                  "expected an integer or a string, got: NimbleOptionsTest"
                ]}
    end
```

My attempt was to return an useful feedback, but of course it breaks the current API. Another option would be to simplify that return to something like `{:error, "A more generic message here"}`.